In [1]:
import numpy as np
import pandas as pd

from matplotlib import pyplot as plt
import seaborn as sns

## Data

This shark attack data comes from the GSAF website ('Global Shark Attack File').

Here is the link: [GSAF](https://www.sharkattackfile.net/investigators.htm)

In [2]:
df = pd.read_excel('data/GSAF5.xls', index_col=0)

In [3]:
df.head()

Date    Year        Type    Country                 Area  \
Case Number                                                                    
2021.05.18   18-May-2021  2021.0  Unprovoked  AUSTRALIA      New South Wales   
2021.05.15   15-May-2021  2021.0  Unprovoked        USA               Hawaii   
2021.05.06   06-May-2021  2021.0  Unprovoked        USA              Florida   
2021.05.03   03-May-2021  2021.0  Unprovoked        USA               Hawaii   
2021.05.01   01-May-2021  2021.0  Unprovoked    JAMAICA  Westmoreland Parish   

                                         Location      Activity  \
Case Number                                                       
2021.05.18                        Turncurry Beach       Surfing   
2021.05.15                Kanaha Beach Park, Maui  Kite Foiling   
2021.05.06   Daytona Beach Shores, Volusia County        Wading   
2021.05.03                Marine Corps Base, Oahu       Surfing   
2021.05.01              Little Bay, Little London  Spearfishing   

                          Name Sex  Age  ...            Species   \
Case Number                              ...                       
2021.05.18    Mark Sanguinetti    M  59  ...  White shark, 4.5 m   
2021.05.15                male    M  35  ...     5 to 8-ff shark   
2021.05.06              female    F  21  ...                 NaN   
2021.05.03   Parker Blanchette    M  14  ...        5.5 ft shark   
2021.05.01     Donovan Haywood    M  53  ...                 NaN   

                           Investigator or Source                         pdf  \
Case Number                                                                     
2021.05.18          S. De Marchi & B. Myatt, GSAF  2021.05.18-Sanguinetti.pdf   
2021.05.15             Star Advertiser, 5/15/2021       2021.05.15-Hawaii.pdf   
2021.05.06   Daytona Beach News-Journal, 5/6/2021      2021.05.05-Daytona.pdf   
2021.05.03        K. McMurray, TrackingSharks.com       2021.05.01-Hawaii.pdf   
2021.05.01        K. McMurray, TrackingSharks.com      2021.05.01-Haywood.pdf   

                                                  href formula  \
Case Number                                                      
2021.05.18   http://sharkattackfile.net/spreadsheets/pdf_di...   
2021.05.15   http://sharkattackfile.net/spreadsheets/pdf_di...   
2021.05.06   http://sharkattackfile.net/spreadsheets/pdf_di...   
2021.05.03   http://sharkattackfile.net/spreadsheets/pdf_di...   
2021.05.01   http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                          href Case Number.1  \
Case Number                                                                    
2021.05.18   http://sharkattackfile.net/spreadsheets/pdf_di...    2021.05.18   
2021.05.15   http://sharkattackfile.net/spreadsheets/pdf_di...    2021.05.15   
2021.05.06   http://sharkattackfile.net/spreadsheets/pdf_di...    2021.05.06   
2021.05.03   http://sharkattackfile.net/spreadsheets/pdf_di...    2021.05.03   
2021.05.01   http://sharkattackfile.net/spreadsheets/pdf_di...    2021.05.01   

            Case Number.2 original order Unnamed: 22 Unnamed: 23  
Case Number                                                       
2021.05.18     2021.05.18         6642.0         NaN         NaN  
2021.05.15     2021.05.15         6641.0         NaN         NaN  
2021.05.06     2021.05.06         6640.0         NaN         NaN  
2021.05.03     2021.05.03         6639.0         NaN         NaN  
2021.05.01     2021.05.01         6638.0         NaN         NaN  

[5 rows x 23 columns]

In [4]:
df.columns

Index(['Date', 'Year', 'Type', 'Country', 'Area', 'Location', 'Activity',
       'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [5]:
df.drop(columns=['pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'], inplace=True)

In [6]:
df.rename(str.lower, axis=1, inplace=True)

In [7]:
df.rename(columns={'fatal (y/n)': 'fatal'}, inplace=True)

In [8]:
df.head()

date    year        type    country                 area  \
Case Number                                                                    
2021.05.18   18-May-2021  2021.0  Unprovoked  AUSTRALIA      New South Wales   
2021.05.15   15-May-2021  2021.0  Unprovoked        USA               Hawaii   
2021.05.06   06-May-2021  2021.0  Unprovoked        USA              Florida   
2021.05.03   03-May-2021  2021.0  Unprovoked        USA               Hawaii   
2021.05.01   01-May-2021  2021.0  Unprovoked    JAMAICA  Westmoreland Parish   

                                         location      activity  \
Case Number                                                       
2021.05.18                        Turncurry Beach       Surfing   
2021.05.15                Kanaha Beach Park, Maui  Kite Foiling   
2021.05.06   Daytona Beach Shores, Volusia County        Wading   
2021.05.03                Marine Corps Base, Oahu       Surfing   
2021.05.01              Little Bay, Little London  Spearfishing   

                          name sex  age                                injury  \
Case Number                                                                     
2021.05.18    Mark Sanguinetti    M  59                                 FATAL   
2021.05.15                male    M  35  2- to 3-inch cuts to posterior thigh   
2021.05.06              female    F  21                          Nick on foot   
2021.05.03   Parker Blanchette    M  14          Lacerations to shin and calf   
2021.05.01     Donovan Haywood    M  53                                 FATAL   

            fatal   time            species   \
Case Number                                    
2021.05.18      Y  11h15  White shark, 4.5 m   
2021.05.15      N  16h45     5 to 8-ff shark   
2021.05.06      N  11h30                 NaN   
2021.05.03      N  16h00        5.5 ft shark   
2021.05.01      Y  08h00                 NaN   

                           investigator or source  
Case Number                                        
2021.05.18          S. De Marchi & B. Myatt, GSAF  
2021.05.15             Star Advertiser, 5/15/2021  
2021.05.06   Daytona Beach News-Journal, 5/6/2021  
2021.05.03        K. McMurray, TrackingSharks.com  
2021.05.01        K. McMurray, TrackingSharks.com

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25808 entries, 2021.05.18 to xx
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   date                    6640 non-null   object 
 1   year                    6639 non-null   float64
 2   type                    6630 non-null   object 
 3   country                 6591 non-null   object 
 4   area                    6173 non-null   object 
 5   location                6092 non-null   object 
 6   activity                6079 non-null   object 
 7   name                    6426 non-null   object 
 8   sex                     6067 non-null   object 
 9   age                     3722 non-null   object 
 10  injury                  6609 non-null   object 
 11  fatal                   6089 non-null   object 
 12  time                    3209 non-null   object 
 13  species                 3660 non-null   object 
 14  investigator or source  6623 non-null

### Data Cleaning

First I will clean and condense the dataframe by removing duplicates. The 19,000 duplicates indicated below are somehow an error in the xls file when read in. The original excel file contains 6641 rows. Trimming off the duplicates leaves a dataframe whos head and tail matches the beginning and end entries of the data opened in excel.

Next, I search for NaN values in the columns I am interested in and remove them. The important columns for me right now are date, year, activity, and location.

In [10]:
df.duplicated().value_counts()

True     19167
False     6641
dtype: int64

In [11]:
df.drop_duplicates(inplace=True)

In [12]:
df.duplicated().value_counts()

False    6641
dtype: int64

In [13]:
df.activity.isna().value_counts()

False    6078
True      563
Name: activity, dtype: int64

In [14]:
df.location.isna().value_counts()

False    6091
True      550
Name: location, dtype: int64

In [26]:
df.dropna(subset=['activity', 'location', 'year', 'fatal', 'country', 'area'], inplace=True)

In [27]:
len(df)

5045

Quickly check below how values for fatal and activity are notated. 

In [17]:
df.fatal.value_counts()

N          4109
Y          1134
UNKNOWN      32
 N            5
2017          1
M             1
Y x 2         1
F             1
Name: fatal, dtype: int64

In [18]:
df.activity.value_counts()

Surfing                                                                                    1015
Swimming                                                                                    781
Fishing                                                                                     407
Spearfishing                                                                                309
Wading                                                                                      152
                                                                                           ... 
Fishing for squid aboard the trawler Shikishima-Maru when the shark leapt into the boat       1
Fishing from dory, shark upset boat & he fell into the water                                  1
Jumping in swells                                                                             1
Chumming for sharks                                                                           1
Fell overboard from the Malacca         

In [19]:
df.year.describe()

count    5284.000000
mean     1941.859008
std       253.640094
min         0.000000
25%      1951.000000
50%      1988.000000
75%      2008.000000
max      2021.000000
Name: year, dtype: float64

Location will be a bit trickier to figure out. Check to see how consistent formatting is and if I will even be able to sort by state or region or whatnot.

In [20]:
df.location.value_counts().index

Index(['New Smyrna Beach, Volusia County', 'Daytona Beach, Volusia County',
       'Ponce Inlet, Volusia County', 'Melbourne Beach, Brevard County',
       'Cocoa Beach, Brevard  County', 'Boa Viagem, Recife',
       'Myrtle Beach, Horry County', 'Isle of Palms, Charleston County',
       'Cocoa Beach, Brevard County',
       'Ponce Inlet, New Smyrna Beach, Volusia County',
       ...
       'Miramar Beach, 46 km south of Mar de Plata, Buenos Aires ',
       'Port Elizabeth Oceanarium', 'New Smyrna Beach, Volusia County  ',
       'Salt Rock', 'Dumbéa', 'Sha’ab Mahmud, Ras Mohamed Park',
       'Artifical reef 3 miles off Manatee Beach, Manatee County',
       'Ocracoke, Lifeguard Beach, National Park Service, Hyde County',
       'Noosa Heads', 'Salter Path, Atlantic Beach, Carteret County'],
      dtype='object', length=3624)

In [21]:
df.country.value_counts().index

Index(['USA', 'AUSTRALIA', 'SOUTH AFRICA', 'NEW ZEALAND', 'PAPUA NEW GUINEA',
       'BRAZIL', 'BAHAMAS', 'MEXICO', 'ITALY', 'REUNION',
       ...
       'NORTH SEA', 'WESTERN SAMOA', 'NETHERLANDS ANTILLES', 'PACIFIC OCEAN ',
       'EQUATORIAL GUINEA / CAMEROON', 'NORWAY', 'DJIBOUTI', 'REUNION ISLAND',
       'SINGAPORE', 'ST KITTS / NEVIS'],
      dtype='object', length=162)

In [22]:
df.area.value_counts().index

Index(['Florida', 'New South Wales', 'Hawaii', 'Queensland', 'California',
       'Western Australia', 'Western Cape Province', 'KwaZulu-Natal',
       'Eastern Cape Province', 'South Carolina',
       ...
       'Bora Bora', 'South Carolina ', '19S, 178?E', 'Majorca',
       'Milne Bay  Province', 'Tuamotu Islands', 'Muala', 'Ralik Chain',
       'Romblon Province', 'Liaoning Province'],
      dtype='object', length=746)

It appears that most 'locations' are given with a county, so using 'area' which is state name for US locations should be more useful. First should check that there arent any rows with an oregon/washington county in 'location' but arent labeled correct in the 'area' column. If there are any, update 'area' to be the state. ( just checked there arent any, every row with an oregon county also properly has 'oregon' in the area column)

I am also noticing now that county names like 'lincoln' might return more locations than just the Oregon county. So, using only the 'area' column by state is the way to go.

In [23]:
counties = ['Clatsop County', 'Tillamook County', 'Lincoln County', 'Lane County', 'Douglas County',
            'Coos County', 'Curry County', 'Clallam County', 'Grays Harbor County', 'Jefferson County',
            'Island County', 'Pacific County', 'Skagit County', 'San Juan County', 'Snohomish County',
            'Wahkiakum County', 'Whatcom County', 'Chehalis County', 'Del Norte County', 'Humboldt County',
            'Mendocino County', 'Sonoma County', 'Marin County']

In [30]:
df.loc[df['location'].str.contains('clatsop', case=False)]

date    year        type country    area  \
Case Number                                                     
2020.12.06.b  06-Dec-2020  2020.0  Unprovoked     USA  Oregon   
2016.10.10    10-Oct-2016  2016.0  Unprovoked     USA  Oregon   
2005.12.24    24-Dec-2005  2005.0  Unprovoked     USA  Oregon   
1988.10.23    23-Oct-1988  1988.0  Unprovoked     USA  Oregon   
1979.11.27    27-Nov-1979  1979.0  Unprovoked     USA  Oregon   

                                                       location  \
Case Number                                                       
2020.12.06.b                       Seaside Cove, Clatsop County   
2016.10.10       Indian Beach, Ecola State Park, Clatsop County   
2005.12.24                       Tillamook Head, Clatsop County   
1988.10.23    Indian Beach, Ecola State Park, just north of ...   
1979.11.27          Haystack Rock, Cannon Beach, Clatsop County   

                                    activity             name sex  age  \
Case Number                                                              
2020.12.06.b                         Surfing  Cole Herrington    M  20   
2016.10.10                           Surfing    Joseph Tanner    M  29   
2005.12.24                           Surfing   Brian Anderson    M  30   
1988.10.23    Surfing (sitting on his board)    Wyndham Kapan    M  21   
1979.11.27                           Surfing      Kenny Doudt    M  20   

                                                     injury fatal   time  \
Case Number                                                                
2020.12.06.b  Non-life-threatening to left lower leg & foot     N  15h20   
2016.10.10              Wounds to upper thigh and lower leg     N  16h00   
2005.12.24                      Lacerations to ankle & calf     N  12h00   
1988.10.23                     Leg bitten & femur fractured     N  17h30   
1979.11.27                          Multiple major Injuries     N  10h20   

                                            species   \
Case Number                                            
2020.12.06.b                                     NaN   
2016.10.10                                       NaN   
2005.12.24                               White shark   
1988.10.23    5.5 m to 6 m [18' to 20']  white shark   
1979.11.27                   White shark, 4 m [13']    

                                         investigator or source  
Case Number                                                      
2020.12.06.b  R. Collier, GSAF & K. McMurray, TrackingSharks...  
2016.10.10                                UP Beacon, 10/12/2016  
2005.12.24             R. Collier, GSAF; L.A. Times, 12/25/2005  
1988.10.23                               R. Collier, pp.104-106  
1979.11.27    D. Miller & R. Collier; R. Collier, p. 76-79; ...

In [33]:
df.loc[df['area'].str.contains('oregon', case=False) | df['area'].str.contains('washington', case=False)]

date    year        type country        area  \
Case Number                                                         
2020.12.06.b  06-Dec-2020  2020.0  Unprovoked     USA      Oregon   
2019.03.05.b  05-Mar-2019  2019.0  Unprovoked     USA      Oregon   
2017.07.14.b  14-Jul-2017  2017.0  Unprovoked     USA  Washington   
2016.10.10    10-Oct-2016  2016.0  Unprovoked     USA      Oregon   
2013.11.22    22-Nov-2013  2013.0  Unprovoked     USA      Oregon   
2012.01.13    13-Jan-2012  2012.0  Unprovoked     USA      Oregon   
2011.12.06    06-Dec-2011  2011.0  Unprovoked     USA      Oregon   
2011.10.20    20-Oct-2011  2011.0  Unprovoked     USA      Oregon   
2011.10.10    10-Oct-2011  2011.0  Unprovoked     USA      Oregon   
2010.10.28    28-Oct-2010  2010.0  Unprovoked     USA      Oregon   
2010.09.27    27-Sep-2010  2010.0  Unprovoked     USA      Oregon   
2006.10.31    31-Oct-2006  2006.0  Unprovoked     USA      Oregon   
2006.08.29.b  29-Aug-2006  2006.0  Unprovoked     USA      Oregon   
2006.07.31.a  31-Jul-2006  2006.0  Unprovoked     USA      Oregon   
2005.12.24    24-Dec-2005  2005.0  Unprovoked     USA      Oregon   
2004.09.20    20-Sep-2004  2004.0  Unprovoked     USA      Oregon   
2002.09.21.b  21-Sep-2002  2002.0  Unprovoked     USA      Oregon   
1998.11.05    05-Nov-1998  1998.0  Unprovoked     USA      Oregon   
1998.04.21    21-Apr-1998  1998.0  Unprovoked     USA      Oregon   
1994.09.21    21-Sep-1994  1994.0  Unprovoked     USA      Oregon   
1993.01.02    02-Jan-1993  1993.0  Unprovoked     USA      Oregon   
1992.09.13    13-Sep-1992  1992.0  Unprovoked     USA      Oregon   
1992.03.08.a  08-Mar-1992  1992.0  Unprovoked     USA      Oregon   
1991.02.24    24-Feb-1991  1991.0  Unprovoked     USA      Oregon   
1989.04.12    12-Apr-1989  1989.0  Unprovoked     USA  Washington   
1988.10.23    23-Oct-1988  1988.0  Unprovoked     USA      Oregon   
1984.09.30.b  30-Sep-1984  1984.0  Unprovoked     USA      Oregon   
1983.08.20.b  20-Aug-1983  1983.0  Unprovoked     USA      Oregon   
1980.10.27    27-Oct-1980  1980.0  Unprovoked     USA      Oregon   
1979.11.27    27-Nov-1979  1979.0  Unprovoked     USA      Oregon   
1976.08.24    24-Aug-1976  1976.0  Unprovoked     USA      Oregon   
1974.09.00       Sep-1974  1974.0  Unprovoked     USA      Oregon   

                                                       location  \
Case Number                                                       
2020.12.06.b                       Seaside Cove, Clatsop County   
2019.03.05.b                     Cape Kiwanda, Tillamook County   
2017.07.14.b         South Beach, Westport, Grays Harbor County   
2016.10.10       Indian Beach, Ecola State Park, Clatsop County   
2013.11.22                       Gleneden Beach, Lincoln County   
2012.01.13                         Lincoln City, Lincoln County   
2011.12.06                                         Seaside Cove   
2011.10.20                              Newport, Lincoln County   
2011.10.10                                              Seaside   
2010.10.28                                             Florence   
2010.09.27                                       Winchester Bay   
2006.10.31                   Siletz River mouth, Lincoln County   
2006.08.29.b                              Florence, Lane County   
2006.07.31.a                                  Oswald State Park   
2005.12.24                       Tillamook Head, Clatsop County   
2004.09.20                                          Gold  Beach   
2002.09.21.b                                       Cape Kiwanda   
1998.11.05                                       Winchester Bay   
1998.04.21                                       Gleneden Beach   
1994.09.21             Short Sand Beach, Oswald West State Park   
1993.01.02                        Bastendorf Beach, Coos County   
1992.09.13                             Gold Beach, Curry County   
1992.03.08.a                     Winchester Bay, Douglas County   

 Clatsop County, Tillamook County, Lincoln County, Lane County, Douglas County, Coos County, Curry County
 
 Clallam County, Grays Harbor County, Jefferson County, Island County, Pacific County, Skagit County, San Juan County, Snohomish County, Wahkiakum County, Whatcom County, Chehalis County
 
 Del Norte County, Humboldt County, Mendocino County, Sonoma County, Marin County
 
 North Coast, Central Coast

### Sorting and subsetting

Now I will look at the subsets and values contained in the Location and Activity columns and choose my area of interest 

In [34]:
pnw_df = df.loc[df['area'].str.contains('oregon', case=False) | df['area'].str.contains('washington', case=False)]

An important decision is the time frame of the data we want to focus on. I could use a single year to estimate the prob I am attacked this year. Or I could look more generally at a lifetime of surfing in the Pacific Northwest. I will opt for the latter. Rather than an average lifespan, I'll choose 50 years as a good length of time one might consistently be able to surf or paddle out in their life.

In [43]:
pnw_df = pnw_df[pnw_df['year'] > 1971]

In [47]:
all_recent_attacks = df[df['year'] > 1971]
len(all_recent_attacks)

3033

### Bayes

In [100]:
def bayes_prob(likelihood, prior, evidence):
    return (likelihood * prior) / evidence

In [184]:
# likelihood is P(PNW surfer | attacked)
# rate of pnw surfers in all attacks

likelihood = (len(pnw_df)/len(all_recent_attacks))
likelihood

0.010550609957138147

In [185]:
# prior is P(attacked)
# total number of attacks over 50yr span divided by the average earth population over those years

# Because our data ranges over the past 50 years, the population we look at should be the average for that time
# The population every 5 years starting in 1971 has been recorded and the average is taken

populations_5yr = [3775759617, 4154666864, 4536996762, 4960567912, 5414289444, 5824891951, 
            6222627000, 6623518000, 7041194000, 7464022000, 7795000000]
av_global_pop = np.mean(populations_5yr)

prior = (len(all_recent_attacks)/av_global_pop)

In [186]:
# data on the population of the population of the PNW in the past 50 years is more limited than global
# Here, I use the population of the PNW in the year 2006 in place of the 50 year average, they should be fairly similar
av_pnw_pop = 6371000+3676000+1464000

# surfertoday magazine has a good article on the number of surfers globally.
# they say it ranges from 13 million to 35 million, but their best guess and most referred to number is 
# 23 million, which I will use here
surfer_pop = 2300000

In [187]:
(2300000**(1/15))/(1.12**(1/15))

2.6353238202357976

In [188]:
# evidence is P(pnw surfer)
# total number of pnw surfers divided by earth population
# will use joint probability to calculate

evidence_surf_pnw = (surfer_pop/av_global_pop)*(av_pnw_pop/av_global_pop)
evidence_pnw = (av_pnw_pop/av_global_pop)
evidence

7.866846446976329e-07

In [217]:
(1/(bayes_prob(likelihood/50, prior/50, evidence_surf_pnw)/50))

17827104.747046243

In [195]:
(1/bayes_prob(likelihood, prior, evidence_pnw))

359718.74999999994

Well I hope that's wrong...

1 in 142 is a gigantic chance of being attacked by a shark while surfing.....
But I think we are overestimating becasue of including the surfing activity in our evidence. It just so happens that all recorded attacks in the PNW are on surfers (probably a result of the fact people dont swim or go scuba diving out here, not that sharks think surfers taste better). Our likelihood would be the exact same if we were considering just PNW attacks or PNW+surfer attacks. 

The intersection of being a PNW resident and surfer in the calculation of the evidence is too specific. Defining someone as a 'surfer' is difficult. Maybe a better idea would be to more generally look at the probability of going surfing in the pacific northwest. This is very hard to calculate though, as evidence usually is. In order to not be reporting such extreme results, I think I will restrict the evidence at hand to just the probability of being in the pacific northwest. Anyone out here could end up going and trying it out I guess. As soon as we add in the joint prob of being a 'surfer' and living here, the posterior skyrockets. As it should though, the more of a true 'surfer' you are, the more  you are in the water.

Also, living in the PNW and being a surfer should be independent events for this to work. But that might not actually be true. Living on a coastline probably makes that more likely. So, I will throw out the way I used intersectional probability in the evidence and stick to asking just about people who live in the pacific northwest.

In [205]:
# probability you are attacked in your life using average yearly attacks
(1/(bayes_prob(likelihood/50, prior/50, evidence_pnw)))/50

# should the evidence also be cut by a factor of 50 here? 

17985937.5

In [208]:
# probability you are attacked in your life using the total attacks over 50 years
(1/bayes_prob(likelihood, prior, evidence_pnw))

359718.74999999994

#### Note for readers:

Apologies for unorganized the bottom of this notebook is. Much copy pasting and updating of evidence happened above. To clarify, the stats I used for the medium article are the results of the two above cells. 

Also, I still included the 1 in 7100 chance in the article as an error to address and give an example of how hard it sometimes is to figure out what the evidence is.

### Fun Facts

1. All recorded attacks in PNW have been on surfers. None have been fatal.
2. All recorded attacks, except 1, north of Marin County have been Great White Sharks.